core/src/test/java/org/elasticsearch/action/bulk/BulkItemResponseTests.java

```diff
@@ -19,17 +19,146 @@
+            int id = randomIntBetween(0, 100);
+            boolean humanReadable = randomBoolean();
+            BytesReference originalBytes = null;
+            BulkItemResponse expectedBulkItemResponse = null;
+
+            if (opType == DocWriteRequest.OpType.INDEX || opType == DocWriteRequest.OpType.CREATE) {
+                expectedBulkItemResponse = new BulkItemResponse(id, opType, randomIndexResponse());
+                originalBytes = toXContent(expectedBulkItemResponse, xContentType, humanReadable);
+
+            } else if (opType == DocWriteRequest.OpType.DELETE) {
+                expectedBulkItemResponse = new BulkItemResponse(id, opType, randomDeleteResponse());
+                originalBytes = toXContent(expectedBulkItemResponse, xContentType, humanReadable);
+
+            } else if (opType == DocWriteRequest.OpType.UPDATE) {
+                Tuple<UpdateResponse, UpdateResponse> updates = randomUpdateResponse(xContentType);
+                expectedBulkItemResponse = new BulkItemResponse(id, opType, updates.v2());
+                originalBytes = toXContent(new BulkItemResponse(id, opType, updates.v1()), xContentType, humanReadable);
+
+            } else {
+                fail("Test does not support opType [" + opType + "]");
+            }
+
+            // Shuffle the XContent fields
+            if (randomBoolean()) {
+                try (XContentParser parser = createParser(xContentType.xContent(), originalBytes)) {
+                    originalBytes = shuffleXContent(parser, randomBoolean()).bytes();
+                }
+            }
+
+            BulkItemResponse parsedBulkItemResponse;
+            try (XContentParser parser = createParser(xContentType.xContent(), originalBytes)) {
+                parsedBulkItemResponse = BulkItemResponse.fromXContent(parser);
+                assertNull(parser.nextToken());
+            }
+
+            assertEquals(expectedBulkItemResponse.getIndex(), parsedBulkItemResponse.getIndex());
+            assertEquals(expectedBulkItemResponse.getType(), parsedBulkItemResponse.getType());
+            assertEquals(expectedBulkItemResponse.getId(), parsedBulkItemResponse.getId());
+            assertEquals(expectedBulkItemResponse.getOpType(), parsedBulkItemResponse.getOpType());
+            assertEquals(expectedBulkItemResponse.getVersion(), parsedBulkItemResponse.getVersion());
+
+            BytesReference finalBytes = toXContent(parsedBulkItemResponse, xContentType, humanReadable);
+            try (XContentParser parser = createParser(xContentType.xContent(), finalBytes)) {
+                assertEquals(XContentParser.Token.START_OBJECT, parser.nextToken());
+                assertEquals(XContentParser.Token.FIELD_NAME, parser.nextToken());
+                assertEquals(opType.getLowercase(), parser.currentName());
+                assertEquals(XContentParser.Token.START_OBJECT, parser.nextToken());
+
+                Map<String, Object> map = parser.map();
```

Comment:
The reason is that `DocWriteResponse` are generated in a randomized fashion and can have `ShardInfo.Failure`. These failures can contain exceptions, which are not directly comparable (because exception lose their types when they are parsed back).

I agree it's complicated in the tests. I don't like the actual map and all DocWriteResponse tests should provide a `Tuple<IndexReponse, IndexResponse> createTestItem()` method that provides the object to serialize and the expected result. It is not the case right now, and the `BulkItemResponseTests` just reuse the current methods from others testing class.

Maybe it could go as it is for now using what other tests already use, and I can clean this up in a follow up PR?